drivers/can/can_nxp_s32_canxl.c

```diff
@@ -0,0 +1,1020 @@
+#endif
+		LOG_ERR("unsupported mode: 0x%08x", mode);
+		return -ENOTSUP;
+	}
+
+	if ((mode & (CAN_MODE_LOOPBACK | CAN_MODE_LISTENONLY))
+				== (CAN_MODE_LOOPBACK | CAN_MODE_LISTENONLY)) {
+		LOG_ERR("unsupported both modes: 0x%08x", mode);
+		return -ENOTSUP;
+	}
+
+	canfd = !!(mode & CAN_MODE_FD);
+	brs = canfd ? true : false;
+
+	if (mode & CAN_MODE_LISTENONLY) {
+		can_nxp_s32_mode = CANEXCEL_LISTEN_ONLY_MODE;
+	} else if (mode & CAN_MODE_LOOPBACK) {
+		can_nxp_s32_mode = CANEXCEL_LOOPBACK_MODE;
+	}
+
+	Canexcel_Ip_EnterFreezeMode(config->instance);
+
+	CanXL_SetFDEnabled(config->base_sic, canfd, brs);
+
+	CanXL_SetOperationMode(config->base_sic, can_nxp_s32_mode);
+
+	Canexcel_Ip_ExitFreezeMode(config->instance);
+
+	return 0;
+}
+
+static int can_nxp_s32_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+
+	__ASSERT(rate != NULL, "Rateptr to get core clock is NULL");
+
+	*rate = config->clock_can;
+
+	return 0;
+}
+
+static int can_nxp_s32_get_max_filters(const struct device *dev, bool ide)
+{
+	ARG_UNUSED(ide);
+
+	return CAN_NXP_S32_MAX_RX;
+}
+
+static int can_nxp_s32_get_max_bitrate(const struct device *dev, uint32_t *max_bitrate)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+
+	*max_bitrate = config->max_bitrate;
+
+	return 0;
+}
+
+static int can_nxp_s32_get_state(const struct device *dev, enum can_state *state,
+						struct can_bus_err_cnt *err_cnt)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+	struct can_nxp_s32_data *data = dev->data;
+	uint32_t sys_status = config->base_sic->SYSS;
+
+	if (state) {
+		if (!data->started) {
+			*state = CAN_STATE_STOPPED;
+		} else {
+			if (sys_status & CANXL_SIC_SYSS_CBOFF_MASK) {
+				*state = CAN_STATE_BUS_OFF;
+			} else if (sys_status & CANXL_SIC_SYSS_CPASERR_MASK) {
+				*state = CAN_STATE_ERROR_PASSIVE;
+			} else if (sys_status & (CANXL_SIC_SYSS_CRXWRN_MASK
+						| CANXL_SIC_SYSS_CTXWRN_MASK)) {
+				*state = CAN_STATE_ERROR_WARNING;
+			} else {
+				*state = CAN_STATE_ERROR_ACTIVE;
+			}
+		}
+	}
+
+	return 0;
+}
+
+static void can_nxp_s32_set_state_change_callback(const struct device *dev,
+							can_state_change_callback_t callback,
+							void *user_data)
+{
+	struct can_nxp_s32_data *data = dev->data;
+
+	data->state_change_cb = callback;
+	data->state_change_cb_data = user_data;
+}
+
+#ifndef CONFIG_CAN_AUTO_BUS_OFF_RECOVERY
+static int can_nxp_s32_recover(const struct device *dev, k_timeout_t timeout)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+	struct can_nxp_s32_data *data = dev->data;
+	enum can_state state;
+	uint64_t start_time;
+	int ret = 0;
+
+	if (!data->started) {
+		return -ENETDOWN;
+	}
+
+	can_nxp_s32_get_state(dev, &state, NULL);
+	if (state != CAN_STATE_BUS_OFF) {
+		return 0;
+	}
+
+	start_time = k_uptime_ticks();
+	config->base_sic->BCFG1 &= (~CANXL_SIC_BCFG1_ABRDIS_MASK);
+
+	if (!K_TIMEOUT_EQ(timeout, K_NO_WAIT)) {
```

Comment:
Not sure I understand this part. If the caller specifies `K_NO_WAIT` as timeout, we do not return whether the bus was recovered but always assume it was?

Reply:
here after the caller specifies `K_NO_WAIT `as timeout. if state is still bus-off, continue to get state until over timeout.
it return `-EAGAIN`.